crates/subspace-farmer/src/piece_cache/tests.rs

```diff
@@ -312,19 +312,23 @@ async fn basic() {
                 },
             };
 
-            archived_segment_headers_sender
-                .send(segment_header)
-                .await
-                .unwrap();
-
-            // Wait for acknowledgement
-            assert_eq!(
-                acknowledge_archived_segment_header_receiver
-                    .next()
+            // Send twice because acknowledgement arrives early, sending twice doesn't have side effects, but ensures
+            // things were processed fully
```

Comment:
I'm also confused with this comment. Since the ack will return the segment_index whether it will process it or already process, why do we need to send it twice though

Reply:
In the rest we are checking the result of cache population. Since we acknowledge before we actually persisted pieces in cache, our checks will be racy. By sending the segment header again we are waiting for piece cache to persist all of the pieces and loop again to get to the next segment header notification. If we get another acknowledgement we deterministically know that the segment was processed fully.